### PR TITLE
FIx violations and enable --warn=all

### PR DESCRIPTION
**Pull request by bongbui321**
**Description** [](A description of the refactor, including the goals it accomplishes.)
For [#29500](https://github.com/commaai/openpilot/issues/29500), Fix violations and enable warn=all

### CODE REVIEW

File: .github/workflows/docs.yaml [hunk 1]
```diff
@@ -29,7 +29,7 @@ jobs:
     - uses: ./.github/workflows/setup-with-retry
     - name: Build openpilot
       run: |
-        ${{ env.RUN }} "scons -j$(nproc)"
+        ${{ env.RUN }} "scons -j$(nproc) --warn=all"
```
Comment (adeebshihadeh) [inline .github/workflows/docs.yaml:32]: can we enable `--warn=all` as the default setting in the SConstruct instead?

Comment (bongbui321) [reply .github/workflows/docs.yaml]: Changed

File: .github/workflows/selfdrive_tests.yaml [hunk 2]
```diff
@@ -344,7 +344,6 @@ jobs:
       - uses: actions/checkout@v4
         with:
           submodules: true
-          ref: ${{ github.event.pull_request.base.ref }}
```
Comment (adeebshihadeh) [inline .github/workflows/selfdrive_tests.yaml:347]: why was this removed? it's needed for this job

File: SConstruct [hunk 3]
```diff
@@ -339,6 +341,29 @@ if GetOption("clazy"):
 
 Export('env', 'qt_env', 'arch', 'real_arch')
 
+def exist_scan(node, env, path):
```
Comment (adeebshihadeh) [inline SConstruct:344]: what's this for?

File: .github/workflows/selfdrive_tests.yaml (hunk 2 above)
Comment (bongbui321) [reply .github/workflows/selfdrive_tests.yaml:347]: I removed it earlier because I set --warn=all explicitly in the build openpilot step of that job and it didn't work since it would pull the Sconsctruct from the base.ref (Sconstruct of master), which didn't work for explicit warn=all setting. But I can add it back now, and it should work normally. doing it now.

File: SConstruct (hunk 3 above)
Comment (bongbui321) [reply SConstruct:344]: Default scanner of Sconstruct filters implicit dependencies by reading the content of the files by regexing "#include" terms. For instance, in the file body/board/inc/stm32f4xx.h, the default Scanner would just filtering based on the "#include" and therefore it would require the file stm32f405xx.h as a dependency. As the pictures below.
image
Also, some of the C standard files would also be required such as "stdint.h" which we generally don't care about since it rarely or never explicitedly edited, and we only care about the dependency files which are in relative folders

**Review by adeebshihadeh — CHANGES_REQUESTED** [review]
Is this really the source of the remaining warnings? This is a pretty core part of SCons, and we shouldn't be doing this scanning ourselves.

Is there a cleaner way to properly fix this? Maybe we can filter these warnings instead?

**Comment by adeebshihadeh** [comment]
> Works only with the assumption that every important implicit dependency is within the repo

Unfortunately, we rely on that behavior, e.g. when we update the Python version.

Thanks for tracking all this stuff down. As is the case sometimes with the bounties, this one doesn't seem feasible to ship right now. Let's just merge the "SetOption('warn', 'all')" but commented out and a link to the issue you linked above it, then we can revisit once SCons addresses this upstream. Once that's done, consider this bounty solved!

**Comment by adeebshihadeh** [comment]
@bongbui321 want to try one of the bounties here? https://github.com/commaai/panda/issues/1770 

should be relatively straightforward and if you want, there's lots of bounties to be made from fixing these https://github.com/commaai/panda/pull/926/files#diff-999690623aab48a8f875013ed611bf2726efc87b076bbbd09ad7bb6d6dbeaec7R15

**Comment by bongbui321** [comment]
Will have a try once I have the chance
